File: ctci/chapter_1_arrays_strings/c/1.9-string_rotation.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* O(n) solution */

/**
 * Determines if a string is a rotation of another string.
 */
int isRotation(char *str1, char *str2)
{
    int len1, len2, i = 0, j = 0;
    char *str1str1;

    len1 = strlen(str1);
    len2 = strlen(str2);
    if (len1 == len2 && len1)
    {
        str1str1 = malloc((len1 * 2) + 1);
        while (str1[i])
        {
            str1str1[j] = str1[i];
            i++;
            j++;
        }
        i = 0;
        while (str1[i])
        {
            str1str1[j] = str1[i];
            i++;
            j++;
        }
        str1str1[j] = '\0';
        /* Check if str2 is a substring is str1str1. */
        return strstr(str1str1, str2)
                   ? 1
                   : 0;
    }
    return 0;
}

int main(void)
{
    printf("%i\n", isRotation("erbottlewat", "waterbottle")); // 1
    printf("%i\n", isRotation("erbottlewa", "waterbottle"));  // 0
    printf("%i\n", isRotation("bot", "tbo"));                 // 1
    printf("%i\n", isRotation("bot", "tob"));                 // 0

    return 0;
}

File: ctci/chapter_1_arrays_strings/1.9-string_rotation.py
# O(n) solution


def is_rotation(str1: str, str2: str) -> bool:
    """
    Determines if a string is a rotation of another string.
    """
    if len(str1) == len(str2) and len(str1):
        str1str1 = str1 * 2
        return str2 in str1str1
    return False


if __name__ == "__main__":
    print(is_rotation("erbottlewat", "waterbottle"))  # true
    print(is_rotation("erbottlewa", "waterbottle"))  # false
    print(is_rotation("bot", "tbo"))  # true
    print(is_rotation("bot", "tob"))  # false
